src/ORM/FieldType/DBGeneratedColumn.php

```diff
@@ -0,0 +1,213 @@
+ * This can be stored (generated whenever columns in the generation statement are updated), or virtual (generated on the fly when accessed).
+ */
+class DBGeneratedColumn extends DBField
+{
+    public const string GENERATION_STORED = 'STORED';
+
+    public const string GENERATION_VIRTUAL = 'VIRTUAL';
+
+    private DBField $childField;
+
+    private string $generationExpression;
+
+    private string $generationType;
+
+    public function __construct(?string $name = null, string $fieldSpec = '', string $expression = '', string $type = DBGeneratedColumn::GENERATION_VIRTUAL)
+    {
+        if (!in_array($type, [DBGeneratedColumn::GENERATION_STORED, DBGeneratedColumn::GENERATION_VIRTUAL])) {
+            throw new InvalidArgumentException('$type must be "STORED" or "VIRTUAL"');
+        }
+        if ($name === null && $fieldSpec === '' && $expression === '') {
+            // DataObjectSchema needs to be able to make a singleton with no args.
+            // This means we need these to be able to be empty....
+            // In that case, we can just skip all the fancy stuff.
+            // Don't use the parent constructor because it calls setValue() even though we have no value.
+            return;
+        } elseif ($fieldSpec === '' || $expression === '') {
+            throw new InvalidArgumentException('$fieldSpec and $expression must not be empty');
+        }
+        $this->generationExpression = $expression;
+        $this->generationType = $type;
+        $this->childField = DBField::create_field($fieldSpec, null, $name);
+        if (!$this->childField->hasMethod('getFieldSpec')) {
+            $class = get_class($this->childField);
+            throw new InvalidArgumentException("Cannot create a generated field based on class '$class' - it needs a 'getFieldSpec' method.");
+        }
+        $this->setFailover($this->childField);
+        parent::__construct($name, []);
+    }
+
+    public function getChildField(): DBField
+    {
+        return $this->childField;
+    }
+
+    public function requireField(): void
+    {
+        $spec = $this->getChildField()->getFieldSpec();
+        if (is_string($spec)) {
+            $defaultValue = $this->getChildField()->getDefaultValue();
+            $spec = DB::get_schema()->makeGenerated(
+                $spec,
+                ['default' => $defaultValue],
+                $this->generationExpression,
+                $this->generationType
+            );
+        } else {
+            $spec['generated'] = [
+                'type' => $this->generationType,
+                'expression' => $this->generationExpression,
+            ];
+        }
+        DB::require_field($this->getTable(), $this->getName(), $spec);
+    }
+
+    public function scaffoldFormField(?string $title = null, array $params = []): ?FormField
+    {
+        return $this->getChildField()->scaffoldFormField($title, $params)->performReadonlyTransformation();
+    }
```

Comment:
Note that most methods in this class are just passing through to the child field.
This means that the data gets treated as though the field was an instance of the child field (e.g. `DBVarchar`) for most purposes. It should be usable in templates the same way, in gridfields, etc.

The main differences are:
1. `scaffoldFormField()` makes the field readonly (you can't save a value into a generated column)
2. `requireField()` obviously adds some extra data (generation storage type and expression)